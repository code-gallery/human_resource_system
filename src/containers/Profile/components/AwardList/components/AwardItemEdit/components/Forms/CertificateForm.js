import React from 'react'
import PropTypes from 'prop-types'
import Select from 'react-select'
import { renderValidation } from 'containers/Profile/utils/validation'
import CustomDatePicker from 'components/FormElements/DatePicker'
import BaseForm from './BaseForm'

class CertificateForm extends BaseForm {
  render() {
    const {
      date,
      description,
      name,
      organisation,
      grade,
      invalidFields
    } = this.props
    return (
      <div>
        <div className="form-group">
          <label className={renderValidation(invalidFields, 'name')}>Name</label>
          <input type="text" name="name" defaultValue={name} onChange={v => this.onTextChange('name', v)} />
        </div>
        <div className="form-group active">
          <label className={renderValidation(invalidFields, 'organisation')}>Organisation</label>
          <Select.Async
            loadOptions={this.getOptions}
            autosize={false}
            labelKey="name"
            valueKey="name"
            value={{ name: organisation }}
            onChange={this.selectAsyncChange}
            onBlur={this.selectAsyncBlur}
            onBlurResetsInput={false}
            autoBlur={true}
            optionRenderer={this.optionRenderer}
            noResultsText="Can't find your organisation? Add it manually and weâ€™ll investigate."
          />
        </div>
        <div className="form-group">
          <label className={renderValidation(invalidFields, 'description')}>Description</label>
          <input
            type="text"
            name="description"
            defaultValue={description}
            onChange={v => this.onTextChange('description', v)}
          />
        </div>
        <div className="form-group">
          <label className={renderValidation(invalidFields, 'grade')}>Grade</label>
          <input type="text" name="grade" defaultValue={grade} onChange={v => this.onTextChange('grade', v)} />
        </div>
        <div className="row">
          <div className="col-xs-12 col-md-6">
            <CustomDatePicker
              handleDate={this.handleChange}
              invalid={invalidFields.indexOf('date') !== -1}
              date={date}
              title="date"
              label="Date"
            />
          </div>
        </div>
      </div>
    )
  }
}

CertificateForm.propTypes = {
  invalidFields: PropTypes.array,
  organisation: PropTypes.string,
  name: PropTypes.string,
  description: PropTypes.string,
  date: PropTypes.string,
  grade: PropTypes.string,
  onFieldChange: PropTypes.func.isRequired,
  orgRegistered: PropTypes.func.isRequired
}

export default CertificateForm
