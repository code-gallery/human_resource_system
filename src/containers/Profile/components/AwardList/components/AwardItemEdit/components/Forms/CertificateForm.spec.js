import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import CertificateForm from './CertificateForm'

describe('containers/Profile/components/.../CertificateForm', () => {
  const commonProps = {
    onFieldChange: jest.fn(),
    orgRegistered: jest.fn(),
    invalidFields: []
  }

  describe('when all inputs have no values', () => {
    it('renders correctly', () => {
      const props = {
        ...commonProps
      }
      const tree = renderer.create(
        <CertificateForm {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('when "name" and "description" are invalid', () => {
    const props = {
      ...commonProps,
      name: '',
      description: '',
      date: '2016-03-01',
      organisation: 'City of Oxford College',
      invalidFields: [ 'name', 'description' ]
    }

    it('renders correctly', () => {
      const tree = renderer.create(
        <CertificateForm {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('when all inputs have values', () => {
    const props = {
      ...commonProps,
      name: 'Award1',
      description: 'Lorem ipsum',
      date: '2016-03-01',
      organisation: 'City of Oxford College'
    }

    it('renders correctly', () => {
      const tree = renderer.create(
        <CertificateForm {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })

    describe('handleChange', () => {
      it('exists on instance', () => {
        const wrapper = shallow(
          <CertificateForm {...props} />
        )
        expect(typeof wrapper.instance().handleChange)
          .toEqual('function')
      })
    })

    describe('onTextChange', () => {
      describe('when "name" changes', () => {
        it('calls props.onFieldChange correctly', () => {
          const wrapper = shallow(
            <CertificateForm {...props} />
          )
          const event = { target: { name: 'name', value: 'Award12' } }
          props.onFieldChange.mockClear()
          wrapper.find('input[name="name"]').simulate('change', event)
          expect(props.onFieldChange)
            .toHaveBeenCalledWith('name', 'Award12')
        })
      })

      describe('when "description" changes', () => {
        it('calls props.onFieldChange correctly', () => {
          const wrapper = shallow(
            <CertificateForm {...props} />
          )
          const event = { target: { name: 'description', value: 'lorem ipsum' } }
          props.onFieldChange.mockClear()
          wrapper.find('input[name="description"]').simulate('change', event)
          expect(props.onFieldChange)
            .toHaveBeenCalledWith('description', 'lorem ipsum')
        })
      })

      describe('when "grade" changes', () => {
        it('calls props.onFieldChange correctly', () => {
          const wrapper = shallow(
            <CertificateForm {...props} />
          )
          const event = { target: { name: 'grade', value: 'A+' } }
          props.onFieldChange.mockClear()
          wrapper.find('input[name="grade"]').simulate('change', event)
          expect(props.onFieldChange)
            .toHaveBeenCalledWith('grade', 'A+')
        })
      })
    })

    describe('optionRenderer', () => {
      it('exists on instance', () => {
        const wrapper = shallow(
          <CertificateForm {...props} />
        )
        expect(typeof wrapper.instance().optionRenderer)
          .toEqual('function')
      })
    })

    describe('selectAsyncChange', () => {
      it('exists on instance', () => {
        const wrapper = shallow(
          <CertificateForm {...props} />
        )
        expect(typeof wrapper.instance().selectAsyncChange)
          .toEqual('function')
      })
    })

    describe('selectAsyncBlur', () => {
      it('exists on instance', () => {
        const wrapper = shallow(
          <CertificateForm {...props} />
        )
        expect(typeof wrapper.instance().selectAsyncBlur)
          .toEqual('function')
      })
    })
  })
})
