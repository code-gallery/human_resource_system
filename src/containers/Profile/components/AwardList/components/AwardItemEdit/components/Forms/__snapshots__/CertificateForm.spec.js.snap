// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Profile/components/.../CertificateForm when "name" and "description" are invalid renders correctly 1`] = `
<div>
  <div
    className="form-group"
  >
    <label
      className="required-error"
    >
      Name
    </label>
    <input
      defaultValue=""
      name="name"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group active"
  >
    <label
      className=""
    >
      Organisation
    </label>
    <div
      className="Select has-value is-clearable is-loading is-searchable Select--single"
      style={undefined}
    >
      <div
        className="Select-control"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <span
          className="Select-multi-value-wrapper"
          id="react-select-3--value"
        >
          <div
            className="Select-value"
            style={undefined}
            title={undefined}
          >
            <span
              aria-selected="true"
              className="Select-value-label"
              id="react-select-3--value-item"
              role="option"
            >
              City of Oxford College
            </span>
          </div>
          <div
            className="Select-input"
            style={
              Object {
                "display": "inline-block",
              }
            }
          >
            <input
              aria-activedescendant="react-select-3--value"
              aria-describedby={undefined}
              aria-expanded="false"
              aria-haspopup="false"
              aria-label={undefined}
              aria-labelledby={undefined}
              aria-owns=""
              className="Select-input"
              id={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              role="combobox"
              tabIndex={undefined}
              value=""
            />
          </div>
        </span>
        <span
          aria-hidden="true"
          className="Select-loading-zone"
        >
          <span
            className="Select-loading"
          />
        </span>
        <span
          className="Select-arrow-zone"
          onMouseDown={[Function]}
        >
          <span
            className="Select-arrow"
            onMouseDown={[Function]}
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className="required-error"
    >
      Description
    </label>
    <input
      defaultValue=""
      name="description"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
    >
      Grade
    </label>
    <input
      defaultValue={undefined}
      name="grade"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="col-xs-12 col-md-6"
    >
      <div
        className="form-group"
      >
        <label
          className=""
        >
          Date
        </label>
        <div
          className="select double-select active"
          onClick={[Function]}
        >
          <span>
            March 2016
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`containers/Profile/components/.../CertificateForm when all inputs have no values renders correctly 1`] = `
<div>
  <div
    className="form-group"
  >
    <label
      className=""
    >
      Name
    </label>
    <input
      defaultValue={undefined}
      name="name"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group active"
  >
    <label
      className=""
    >
      Organisation
    </label>
    <div
      className="Select has-value is-clearable is-loading is-searchable Select--single"
      style={undefined}
    >
      <div
        className="Select-control"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <span
          className="Select-multi-value-wrapper"
          id="react-select-2--value"
        >
          <div
            className="Select-value"
            style={undefined}
            title={undefined}
          >
            <span
              aria-selected="true"
              className="Select-value-label"
              id="react-select-2--value-item"
              role="option"
            />
          </div>
          <div
            className="Select-input"
            style={
              Object {
                "display": "inline-block",
              }
            }
          >
            <input
              aria-activedescendant="react-select-2--value"
              aria-describedby={undefined}
              aria-expanded="false"
              aria-haspopup="false"
              aria-label={undefined}
              aria-labelledby={undefined}
              aria-owns=""
              className="Select-input"
              id={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              role="combobox"
              tabIndex={undefined}
              value=""
            />
          </div>
        </span>
        <span
          aria-hidden="true"
          className="Select-loading-zone"
        >
          <span
            className="Select-loading"
          />
        </span>
        <span
          className="Select-arrow-zone"
          onMouseDown={[Function]}
        >
          <span
            className="Select-arrow"
            onMouseDown={[Function]}
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
    >
      Description
    </label>
    <input
      defaultValue={undefined}
      name="description"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
    >
      Grade
    </label>
    <input
      defaultValue={undefined}
      name="grade"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="col-xs-12 col-md-6"
    >
      <div
        className="form-group"
      >
        <label
          className=""
        >
          Date
        </label>
        <div
          className="select double-select active"
          onClick={[Function]}
        >
          <span>
            April 2018
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`containers/Profile/components/.../CertificateForm when all inputs have values renders correctly 1`] = `
<div>
  <div
    className="form-group"
  >
    <label
      className=""
    >
      Name
    </label>
    <input
      defaultValue="Award1"
      name="name"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group active"
  >
    <label
      className=""
    >
      Organisation
    </label>
    <div
      className="Select has-value is-clearable is-loading is-searchable Select--single"
      style={undefined}
    >
      <div
        className="Select-control"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <span
          className="Select-multi-value-wrapper"
          id="react-select-4--value"
        >
          <div
            className="Select-value"
            style={undefined}
            title={undefined}
          >
            <span
              aria-selected="true"
              className="Select-value-label"
              id="react-select-4--value-item"
              role="option"
            >
              City of Oxford College
            </span>
          </div>
          <div
            className="Select-input"
            style={
              Object {
                "display": "inline-block",
              }
            }
          >
            <input
              aria-activedescendant="react-select-4--value"
              aria-describedby={undefined}
              aria-expanded="false"
              aria-haspopup="false"
              aria-label={undefined}
              aria-labelledby={undefined}
              aria-owns=""
              className="Select-input"
              id={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              role="combobox"
              tabIndex={undefined}
              value=""
            />
          </div>
        </span>
        <span
          aria-hidden="true"
          className="Select-loading-zone"
        >
          <span
            className="Select-loading"
          />
        </span>
        <span
          className="Select-arrow-zone"
          onMouseDown={[Function]}
        >
          <span
            className="Select-arrow"
            onMouseDown={[Function]}
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
    >
      Description
    </label>
    <input
      defaultValue="Lorem ipsum"
      name="description"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
    >
      Grade
    </label>
    <input
      defaultValue={undefined}
      name="grade"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="col-xs-12 col-md-6"
    >
      <div
        className="form-group"
      >
        <label
          className=""
        >
          Date
        </label>
        <div
          className="select double-select active"
          onClick={[Function]}
        >
          <span>
            March 2016
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
